### PR TITLE
fix(1839): Stop active build step with build timeout

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

To timeout a build when build container is not available we need the build start time information

## Objective

This PR will read the config from timeoutConfig queue and do the time calculations to timeout a build. After timeout it will stop the active build step and mark the build as failure

## References

https://github.com/screwdriver-cd/screwdriver/issues/1839

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: lib/helper.js [hunk 1]
```diff
@@ -40,6 +40,82 @@ function updateBuildStatus(updateConfig, callback) {
...
+ */
+async function updateStepStop(stepConfig) {
+    const { redisInstance, buildId, stepName, code } = stepConfig;
+    const buildConfig = await redisInstance.hget(`${queuePrefix}buildConfigs`, buildId)
+        .then(JSON.parse);
+
+    // if buildConfig got deleted already, do not update
+    if (!buildConfig) return null;
+
+    const promise = () => new Promise((resolve, reject) => {
```
Comment (tkyi) [inline lib/helper.js]:
```suggestion
    return new Promise((resolve, reject) => {
```

File: lib/helper.js [hunk 2]
```diff
@@ -40,6 +40,82 @@ function updateBuildStatus(updateConfig, callback) {
...
+ */
+async function getCurrentStep(config) {
+    const { redisInstance, buildId } = config;
+    const buildConfig = await redisInstance.hget(`${queuePrefix}buildConfigs`, buildId)
+        .then(JSON.parse);
+
+    // if buildConfig got deleted already, do not update
+    if (!buildConfig) return null;
+
+    const promise = () => new Promise((resolve, reject) => {
```
Comment (tkyi) [inline lib/helper.js]:
```suggestion
    return new Promise((resolve, reject) => {
```

File: lib/helper.js [hunk 3]
```diff
@@ -40,6 +40,82 @@ function updateBuildStatus(updateConfig, callback) {
         });
 }
 
+/**
+ * Updates the step with code and end time
+ * @param {Object} stepConfig
+ * @param {Object} redisInstance
```
Comment (tkyi) [inline lib/helper.js]:
```suggestion
 * @param {Object} stepConfig.redisInstance             Redis instance
```

File: lib/helper.js [hunk 4]
```diff
@@ -40,6 +40,82 @@ function updateBuildStatus(updateConfig, callback) {
...
+ * @param {String} buildId
+ * @param {String} stepName
+ * @param {Integer} code
+ */
```
Comment (tkyi) [inline lib/helper.js]: */ @return {Promise}                                     Updated step

File: lib/helper.js [hunk 5]
```diff
@@ -40,6 +40,82 @@ function updateBuildStatus(updateConfig, callback) {
...
+            return reject(err);
+        });
+    });
+
+    return promise();
+}
+
+/**
+ * Gets the current active step in the build
+ * @param {Object} redisInstance
```
Comment (tkyi) [inline lib/helper.js]:
```suggestion
 * @param {Object} config.redisInstance          Redis instance
```

File: lib/helper.js [hunk 6]
```diff
@@ -40,6 +40,82 @@ function updateBuildStatus(updateConfig, callback) {
...
+ * @param {String} buildId
+ * @param {Object} config
+ */
```
Comment (tkyi) [inline lib/helper.js]:
```suggestion
 * @return {Promise}                               Active step
```

File: lib/timeout.js [hunk 7]
```diff
@@ -1,80 +1,107 @@
...
-const DEFAULT_BUILD_TIMEOUT = 60;
+const TIMEOUT_CODE = 3;
 
 /**
- * Check if the build has timed out
- * If yes, abort build.
- * @method check
- * @return {Promise}
+ * Wrapper function to process timeout logic
+ * @param {Object} timeoutConfig
```
Comment (tkyi) [inline lib/timeout.js]: missing descriptions, method name, return value

File: lib/timeout.js [hunk 8]
```diff
@@ -1,80 +1,107 @@
...
+        await redis.hdel(`${queuePrefix}timeoutConfigs`, buildId);
+
+        logger.error(`Error occured while checking timeout ${err}`);
+    }
+}
+
+/**
+ * Check if the build has timed out
+ * If yes, abort build.
+ * @method check
```
Comment (tkyi) [inline lib/timeout.js]: missing param

File: lib/timeout.js [hunk 9]
```diff
@@ -1,80 +1,107 @@
...
-            const startTime = buildConfig.enqueueTime;
-
-            if (!startTime) {
-                // there is no enqueue time set for the build
-                logger.warn(`EnqueueTime not set for buildId: ${buildId}`);
+            if (!step) {
+                // there is no startTime set for the build
+                logger.error(`No active step found for  ${buildId}`);
 
                 return;
```
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply lib/timeout.js]: correct. will set build to failure even if there is not active step

File: test/timeout.test.js [hunk 10]
```diff
@@ -79,33 +84,36 @@ describe('Timeout test', () => {
...
 
-                // helperMock.updateBuildStatus.yieldsAsync(null, {});
-
                 const buildId = '333';
                 const buildConfig = {
                     jobId: 2,
                     jobName: 'deploy',
                     annotations: {
-                        timeout: '50'
+                        'scredriver.cd/timeout': 50
```
Comment (tkyi) [inline test/timeout.test.js]:
```suggestion
                        'screwdriver.cd/timeout': 50
```

File: test/timeout.test.js [hunk 11]
```diff
@@ -115,57 +123,122 @@ describe('Timeout test', () => {
...
-                buildId,
-                eventId: 76,
-                enqueueTime: now
-            };
+                const buildId = '333';
+                const buildConfig = {
+                    jobId: 2,
+                    jobName: 'deploy',
+                    annotations: {
+                        'scredriver.cd/timeout': 50
```
Comment (tkyi) [inline test/timeout.test.js]:
```suggestion
                        'screwdriver.cd/timeout': 50
```

**Pull request merged**